### PR TITLE
Do not resolve original jar via hash after instrumentation in the final buildscript classpath

### PR DESCRIPTION
**Pull request by asodja**
The problem is, that we tried to figure out original jar for the final classpath via hash. Additionally we included also "is in global cache" value as part of a hash value. But between invocations users can change repository from remote to local that has same libraries. And suddenly hash doesn't match anymore, since a jar is now in "non global cache directory" (local cache is not "global cache", since it's outside Gradle user home, but jars from remote are cached in the global cache aka Gradle user home). 

The fix is, that original jar in the final classpath is not retrieved via hash anymore, but we pass original jar through the whole transform chain. We still need to retrieve original jar via hash when we generate instrumented jar inside the transform though. But there we don't care about the location of the jar we instrument, only about content.

Fixes https://github.com/gradle/gradle/issues/28301

Before (left) vs Now (right):
```mermaid
  flowchart TB
  analysis{{"Analysis\n(Artifact transform)"}}
  mergeAnalysis{{"Merge analysis\n(Artifact transform)"}}
  instrument{{"Instrument and upgrade\n(Artifact transform)\nWe find original Jar via file hash\nWe cannot output original jar here unless we copy it,\nsince it's not an input to this transform."}}
  originalJar[[Original jar]]
  analysisMetadata[["Analysis metadata\n(original file hash + analysis)"]]
  mergeMetadata[["Merged analysis metadata\n(original file hash + merged analysis)"]]
  instrumentedJar[[Instrumented Jar]]
  fileWithHashToOriginalJar[[A copy of original Jar\nor a file with original Jar hash]]
  logic{{"Logic that resolves original jar via hash\n(outside the artifact transform)"}}
  originalJarOrCopy[[A copy of original Jar\nor original Jar]]
  outputClasspath[[Output classpath]]
  
  originalJar --> analysis --> analysisMetadata --> mergeAnalysis --> mergeMetadata --> instrument
  instrument --> instrumentedJar
  instrument --> fileWithHashToOriginalJar --> logic --> originalJarOrCopy
  instrumentedJar --> outputClasspath
  originalJarOrCopy --> outputClasspath
  
  analysis2{{"Analysis\n(Artifact transform)"}}
  mergeAnalysis2{{"Merge analysis\n(Artifact transform)"}}
  mergeAnalysis3{{"Merge analysis\n(Artifact transform)\nJar just passes through"}}
  instrument2{{"Instrument and upgrade\n(Artifact transform)\nWe find original Jar via file hash"}}
  instrument3{{"Instrument and upgrade\n(Artifact transform)\nJar is copied or it's passed through"}}
  originalJar2[[Original jar]]
  originalJar3[[Original jar]]
  originalJar4[[Original jar]]
  analysisMetadata2[["Analysis metadata\n(original file hash + analysis)"]]
  mergeMetadata2[["Merged analysis metadata\n(original file hash + merged analysis)"]]
  instrumentedJar2[["Instrumented Jar"]]
  originalJarOrCopy2[[A copy of original Jar\nor original Jar]]
  outputClasspath2[[Output classpath]]
  
  originalJar2 --> analysis2 --> analysisMetadata2 --> mergeAnalysis2 --> mergeMetadata2 --> instrument2
  analysis2 --> originalJar3 --> mergeAnalysis3 --> originalJar4 --> instrument3
  instrument2 --> instrumentedJar2
  instrument3 --> originalJarOrCopy2 
  instrumentedJar2 --> outputClasspath2
  originalJarOrCopy2 --> outputClasspath2
```

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle test test APT

**Comment by asodja** [comment]
@bot-gradle test APT

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test QF

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/utils/InstrumentationClasspathMerger.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+import java.io.File;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class InstrumentationClasspathMerger {
```
Comment (asodja) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/utils/InstrumentationClasspathMerger.java:33]: This class is just some code extracted from DefaultScriptClassPathResolver.

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test APT

**Comment by asodja** [comment]
@bot-gradle clean

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java [hunk 2]
```diff
@@ -163,6 +154,12 @@ private static FileCollection getAnalysisResult(Configuration classpathConfigura
         }).getFiles();
     }
 
+    private static ArtifactView getOriginalDependencies(Configuration classpathConfiguration) {
+        return classpathConfiguration.getIncoming().artifactView((Action<? super ArtifactView.ViewConfiguration>) config -> {
+            config.componentFilter(it -> !isGradleApi(it));
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java:159]: ❓ We did not filter the Gradle API before, is this something we only need now?

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java [hunk 3]
```diff
@@ -79,53 +76,30 @@ public interface Parameters extends TransformParameters {
     @InputArtifact
     public abstract Provider<FileSystemLocation> getInput();
 
-    protected void execute(@Nullable File artifactToInstrument, TransformOutputs outputs, Consumer<TransformOutputs> originalFileProducer) {
+    protected void doTransform(File artifactToTransform, TransformOutputs outputs) {
         createInstrumentationClasspathMarker(outputs);
-        if (artifactToInstrument == null || !artifactToInstrument.exists()) {
-            createNewFile(outputs.file(ORIGINAL_FILE_DOES_NOT_EXIST_MARKER));
+        if (!artifactToTransform.exists()) {
+            GFileUtils.touch(outputs.file(ORIGINAL_FILE_DOES_NOT_EXIST_MARKER));
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:82]: ❓ Why this change?

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java [hunk 4]
```diff
@@ -140,13 +114,25 @@ private static String getOutputPath(File input, Function<String, String> instrum
...
+            // Jars that are already in the global cache don't need to be copied, since
+            // the global caches are additive only and jars shouldn't be deleted or changed during the build.
+            outputs.file(input);
+        } else {
+            // Jars that are in some mutable location (e.g. build/ directory) need to be copied to the global cache,
+            // since daemon keeps them locked when loading them to a classloader, which prevents e.g. deleting the build directory on windows
+            File copyOfOriginalFile = outputs.file(ORIGINAL_DIR_NAME + "/" + input.getName());
+            GFileUtils.copyFile(input, copyOfOriginalFile);
+        }
     }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:132]: 🤔 This looks very brittle, depending on implementation details to make this decision. I wonder how we could make this more future proof...

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/ExternalDependencyInstrumentingArtifactTransform.java [hunk 5]
```diff
@@ -26,39 +26,33 @@
...
-        if (metadata.getArtifactHash().equals(FILE_MISSING_HASH)) {
-            execute(null, outputs, __ -> {});
-        } else {
-            String hash = metadata.getArtifactHash();
-            long contextId = getParameters().getContextId().get();
-            File originalArtifact = getParameters().getBuildService().get().getOriginalFile(contextId, hash);
-            execute(originalArtifact, outputs, __ -> writeOriginalFilePlaceholder(hash, outputs));
+        // We simulate fan-in behaviour:
+        // We expect that a transform before this one outputs two artifacts: 1. analysis metadata and 2. the original file.
+        // So if the input is analysis metadata we use it and create instrumented artifact, otherwise it's original artifact and we output that.
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/ExternalDependencyInstrumentingArtifactTransform.java:42]: ❓ Do I understand correctly that we only create an instrumented artifact if analysis metadata is available, and if not, we pass the uninstrumented artifact through?

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/MergeInstrumentationAnalysisTransform.java [hunk 6]
```diff
@@ -108,16 +117,24 @@ public void transform(TransformOutputs outputs) {
...
         return buildService.getInstrumentationTypeRegistry(contextId);
     }
+
+    private static void doOutputOriginalArtifact(File input, TransformOutputs outputs) {
+        if (input.isDirectory()) {
+            outputs.dir(input);
+        } else {
+            outputs.file(input);
+        }
+    }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/MergeInstrumentationAnalysisTransform.java:139]: 🤔 We should at least move this to some utility class.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/utils/InstrumentationClasspathMerger.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/utils/InstrumentationClasspathMerger.java:33]: The only change is that now we use `List<File>` to avoid deduplication, right?

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java (hunk 2 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java:159]: We don't need to pass Gradle API as input to the artifact transforms. Before we just did a bit more work unnecessary

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java (hunk 3 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:82]: In previous iteration `artifactToInstrument` passed here could be null, but with the change that should not be possible anymore, since we filter unexisting jars out already before that step.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java (hunk 4 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:132]: I would need some more context, what do you mean with `implementation details` here?

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/ExternalDependencyInstrumentingArtifactTransform.java (hunk 5 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/ExternalDependencyInstrumentingArtifactTransform.java:42]: So `analysis metadata` will always be available if jar exist.

So here what happens is:
```mermaid
flowchart TB
previousStep{{"Previous step\n(Artifact transform)"}}
instrumentationAndUpgrade{{"Instrumentation and upgrade\n(Artifact transform)"}} 
instrumentationAndUpgrade2{{"Instrumentation and upgrade\n(Artifact transform)"}} 
originalJar[["Original Jar"]]
originalJar2[["Original Jar"]]
metadata[["Metadata"]]
instrumentedJar[["Instrumented Jar"]]
finalClasspath[["Final classpath"]]

previousStep --> metadata --> instrumentationAndUpgrade --> instrumentedJar --> finalClasspath
previousStep --> originalJar --> instrumentationAndUpgrade2 --> originalJar2 --> finalClasspath
```

Basically previous step outputs two artifacts. And then we have this if:
if `Instrumentation and upgrade` detects that it's a metadata:
    it discovers original Jar via build service and outputs instrumented jar
else:
    it's just original jar and we just output it
 
Alternative approach would be, to not have that `else` case and pass original jar through. But basically we could find the original jar for the instrumented jar from the original classpath after instrumentation (this could also work, but had some quirks when I tried to allign two classpaths, so I went with this)

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/utils/InstrumentationClasspathMerger.java (hunk 1 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/utils/InstrumentationClasspathMerger.java:33]: Yes, and toString methods on classes, but that I used just for debugging and left it here

File: subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathInstrumentationIntegrationTest.groovy [hunk 7]
```diff
@@ -144,6 +151,7 @@ class BuildScriptClasspathInstrumentationIntegrationTest extends AbstractIntegra
 
     def "order of entries in the effective classpath stays the same as in the original classpath"() {
         given:
+        requireOwnGradleUserHomeDir()
```
[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathInstrumentationIntegrationTest.groovy:154]: Good idea! I can add a method with a "because" and deprecate the old method for now. Thinking about just deprecating it, since the method is used in lots of places and would increase the size of a review a lot. 

What do you think about that idea?

[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/classpath/BuildScriptClasspathInstrumentationIntegrationTest.groovy:154]: Added `requireOwnGradleUserHomeDir(String because)` method

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/MergeInstrumentationAnalysisTransform.java (hunk 6 above)
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/MergeInstrumentationAnalysisTransform.java:139]: Moved to instrumentation utils

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/InstrumentationAnalysisTransform.java [hunk 8]
```diff
@@ -162,12 +161,18 @@ private void writeOutput(File artifact, TransformOutputs outputs, Map<String, Se
 
         File dependenciesFile = new File(outputDir, DEPENDENCIES_FILE_NAME);
         serializer.writeTypes(dependenciesFile, dependencies);
+
+        if (artifact.isDirectory()) {
+            outputs.dir(artifact);
+        } else {
+            outputs.file(artifact);
+        }
```
[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/InstrumentationAnalysisTransform.java:169]: Moved to instrumentation utils for now, since `TransformOutputs` is a public interface

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test APT

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test QF

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java (hunk 3 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java:82]: Can we then keep using `createNewFile()`? Using `touch` gives the idea that we expect outputs to sometimes exist here.

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test this APT

**Comment by asodja** [comment]
@bot-gradle test this APT

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/services/CacheInstrumentationDataBuildService.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        private final Lazy<Map<String, File>> hashToOriginalFile;
+        private final Map<File, String> hashCache;
+        private final InjectedInstrumentationServices internalServices;
+
+        @Inject
+        public ResolutionData(InjectedInstrumentationServices internalServices) {
+            this.hashCache = new ConcurrentHashMap<>();
+            this.hashToOriginalFile = Lazy.locking().of(() -> {
+                Set<File> originalClasspath = getOriginalClasspath().getFiles();
+                Map<String, File> originalFiles = new HashMap<>(originalClasspath.size());
```
[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/services/CacheInstrumentationDataBuildService.java:139]: Oh, that is nice, thanks! I didn't know about that. I just knew that something similar exists in Java 17 (or 21?). 
Fixed in https://github.com/gradle/gradle/pull/28514

File: platforms/jvm/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/incremental/serialization/HierarchicalNameSerializerTest.groovy
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/jvm/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/incremental/serialization/HierarchicalNameSerializerTest.groovy:1]: Nice catch, fixed in https://github.com/gradle/gradle/pull/28514

File: platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java [hunk 10]
```diff
@@ -341,7 +355,7 @@ private static boolean areInstrumentedAndOriginalEntriesValid(File instrumentedE
     }
 
     private static boolean isInstrumentedMarkerFile(File classPathEntry) {
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java:357]: Done in https://github.com/gradle/gradle/pull/28514

File: platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java [hunk 11]
```diff
@@ -247,86 +264,83 @@ public String toString() {
...
-        for (int i = 0; i < classPath.getAsFiles().size(); ++i) {
-            File inputFile = classPath.getAsFiles().get(i);
-            if (isInstrumentedMarkerFile(inputFile)) {
+        for (int i = 0; i < classPath.size(); ++i) {
+            File markerFile = classPath.get(i);
+            if (markerFile.getName().equals(AGENT_INSTRUMENTATION_MARKER_FILE_NAME)) {
                 // Marker file will be followed by instrumented entry and original entry.
                 i += 2;
+                ++resultSize;
+            } else if (markerFile.getName().equals(LEGACY_INSTRUMENTATION_MARKER_FILE_NAME)) {
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java:307]: You are right, we can't mix them. 
For compressibility I think  it makes sense we handle both in the same way. And also that way we handle legacy vs non-legacy just here when we resolve final classpath. So I plan to leave this for now.

Feel free to comment if you don't agree.

File: platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java [hunk 12]
```diff
@@ -247,86 +264,83 @@ public String toString() {
...
      */
-    private static int computeResultSizeOfInstrumentingArtifactTransformOutput(ClassPath classPath) {
+    private static int computeResultSizeOfInstrumentingArtifactTransformOutput(List<File> classPath) {
         int resultSize = 0;
-        for (int i = 0; i < classPath.getAsFiles().size(); ++i) {
-            File inputFile = classPath.getAsFiles().get(i);
-            if (isInstrumentedMarkerFile(inputFile)) {
+        for (int i = 0; i < classPath.size(); ++i) {
+            File markerFile = classPath.get(i);
+            if (markerFile.getName().equals(AGENT_INSTRUMENTATION_MARKER_FILE_NAME)) {
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/core-runtime/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java:303]: Yes, unfortunately this project uses Java 6 so I can't use switch on strings. But I changed values to enum in https://github.com/gradle/gradle/pull/28514, so this is addressed there

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/mutator/ClearArtifactTransformCacheWithoutInstrumentedJarsMutator.groovy [hunk 13]
```diff
@@ -25,8 +25,9 @@ import java.nio.file.Files
 import java.nio.file.Path
 import java.nio.file.attribute.BasicFileAttributes
 
-import static org.gradle.internal.classpath.TransformedClassPath.INSTRUMENTED_MARKER_FILE_NAME
-
+import static org.gradle.api.internal.initialization.transform.utils.InstrumentationTransformUtils.ANALYSIS_OUTPUT_DIR
+import static org.gradle.api.internal.initialization.transform.utils.InstrumentationTransformUtils.MERGE_OUTPUT_DIR
+import static org.gradle.internal.classpath.TransformedClassPath.INSTRUMENTATION_CLASSPATH_MARKER_FILE_NAME
 /**
  * A mutator that cleans up the artifact transform cache directory, but leaves folders with the instrumented jars.
```
[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/mutator/ClearArtifactTransformCacheWithoutInstrumentedJarsMutator.groovy:32]: Added some additional information in https://github.com/gradle/gradle/pull/28514, thanks!
